CactusTutorial.ipynb: Add kuibit

Sbozzolo opened the pull request:
As briefly discussed in ET call on Thursday 14 Jan, I added `kuibit` to `CactusTutorial`. 

In the tutorial, I am showing two different ways to use of `kuibit` to plot a timeseries and a grid function. The first is using existing examples, the second is writing your custom Python code, where I also highlight how you can make some basic manipulations. Of course, this barely scratches the surface of what `kuibit can do.

Sbozzolo committed:
CactusTutorial.ipynb: Add kuibit

rhaas80 commented:
@Sbozzolo please create a Cactus ticket at https://bitbucket.org/einsteintoolkit/tickets/issues/new since lone pull requests mostly go unnoticed (too many repositories to monitor them all or check them once one recalls that there was a pull request). Please add "Please review" as a comment to the ticket so that it is found in the ET call using the "Tickets ready for review" standing ET call item.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -738,12 +738,17 @@
-    "# This cell enables inline plotting in the notebook\n",
-    "%matplotlib inline\n",
+    "%%bash\n",
+    "pip3 install -U --user kuibit"
```
Is there a way to install kuibit inside of the Cactus tree? Eg by setting `PYTHONUSERBASE` to Cactus' root directory (the `export` is a dummy in the notebook, but required if one enters the commands at the command line):
```
%%bash
export PYTHONUSERBASE="$PWD"
pip3 install -U --user kuibit
```

and in the notebook setup cell at the beginning:
```
os.environ["PYTHONUSERBASE"] = os.environ["HOME"] + "/Cactus"
```

This way, a user can wipe their Cactus tree and remove all traces of having tried it out.

See https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUSERBASE for docs on this variable.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -729,7 +729,7 @@
     "<h2>Plotting the Output</h2>\n",
     "\n",
-    "The following commands will generate a simple line plot of the data. They will work in a python script as easily as they do in the notebook (just remove the \"%matplotlib inline\" directive)."
+    "The simplest way to analyze a simulation is using [kuibit](https://sbozzolo.github.io/kuibit/), a Python package that has to be installed separately. You can install `kuibit` with `pip`:"
```
sicne you use `pip3` below, it should be `pip3` here as well.

rhaas80 replied:
I would add that kuibit is part of the ET (but has to be installed separately to get all dependencies installed).

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -738,12 +738,17 @@
-    "# This cell enables inline plotting in the notebook\n",
-    "%matplotlib inline\n",
+    "%%bash\n",
+    "pip3 install -U --user kuibit"
```
Is there a danger of `pip3` asking questions of the user (eg overwrite, reinstall, verify certificate)? In that case one should add an option to force non-interactive, yes-to-everything mode since jupyter notebooks do not allow for user input to shell code.

Sbozzolo replied:
I am not sure about that. I don't see `pip3 --force install` option.

rhaas80 replied:
there seems to be:  
\`\`\`  
 --no-input                  Disable prompting for input.  
\`\`\`  
It does ask questions though?

Sbozzolo replied:
In the best case scenario (everything is working as expected), no questions are asked.

rhaas80 replied:
ok, let's give it a try and see if we get any reports by confused users.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -738,12 +738,17 @@
-    "import matplotlib\n",
-    "import numpy as np\n",
-    "import matplotlib.pyplot as plt"
+    "The Einstein Toolkit comes with several [kuibit examples](https://sbozzolo.github.io/kuibit/#examples) that are ready to be used. Most examples are command-line scripts that produce plots. For example, if we want to plot the maximum rest-mass density as a function of time, we can use the `plot_timeseries.py` example:"
```
maybe point to the examples in their "canonical" (ie in einsteintoolkit.th) checkout location.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -753,8 +758,21 @@
+    "# the simulation data lives. We do so to keep this tutorial general.\n",
+    "# Normally you would just input the path or you would run the\n",
+    "# script from the output directory.\n",
+    "datadir=$(dirname $(./simfactory/bin/sim get-output-dir tov_ET))\n",
```
I would not call `datadir` the output directory since the output directory is what is returned by `get-output-dir` and `dirname ...` return the parent, which is the simulation directory.

Sbozzolo replied:
When I tried on my machine `get-output-dir` returned the output dir of the last simulation restart (ie, with `output-XXXX`). In `kuibit`, we do not want that: we want the top-level directory. This is what `dirname` does.

rhaas80 replied:
Oh, I am not saying that would you do is wrong. I assume you tested it. I am trying to say hat you would "run the script from the output directory" and my guess is that you mean you would run from `$datadir` which is not the output directory (in simfactory's terms).  
Ie aiming for consistent names everywhere.

Sbozzolo replied:
I see! That's right!

Sbozzolo replied:
What is the correct wording to say that one should run the script from the top-level folder of a given simulation data?

rhaas80 replied:
I would avoid "output directory". Why not use "data directory" which is the name of variable and the name of the command line option.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -763,21 +781,55 @@
+    "# tutorial general.\n",
     "import os\n",
-    "outdir = os.popen(\"./simfactory/bin/sim get-output-dir tov_ET\").read().rstrip(\"\\n\")\n",
+    "outdir = os.popen(\"./simfactory/bin/sim get-output-dir tov_ET\").read().rstrip(\"\\n\")[:-12]\n",
```
nnono. If you want to do the equivalent of `dirname` please do something like:
```
os.path.join(os.path.split(output_dir)[:-1]))
```
to remove the last path component. `-12` is quite unintuitive and will fail if the simulation name is not 11 characters long.

Sbozzolo replied:
See comment above, `get-output-dir` returns the directory with `output-XXXX`, the -12 removes that.

rhaas80 replied:
I understand that :-). My comment is that this code is not very robust and that a more robust way would be preferable.

Sbozzolo replied:
I defined a helper function to hide all the complexity in the preamble of the notebook.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -763,21 +781,55 @@
+    "# the simulation data lives. We do so to keep this tutorial general.\n",
+    "# Normally you would just input the path or you would run the\n",
+    "# script from the output directory.\n",
+    "datadir=$(dirname $(./simfactory/bin/sim get-output-dir tov_ET))\n",
```
could use an `os.environ[FOO]` trick same as what I suggest above for `PYTHONUSERBASE` to avoid having to redefine dataqdir all the time.

It's also possible to use Python variables in  `!` lines (and in `%%bash` cells since we remap `%%bash` to `!`) using `{foo}` syntax, eg:
```
foo = 12
```
```
# this allows you to use "cd" in cells to change directories instead of requiring "%cd"
%automagic on
# override IPython's default %%bash to not buffer all output
from IPython.core.magic import register_cell_magic
@register_cell_magic
def bash(line, cell): get_ipython().system(cell)
```
```
%%bash
echo {foo}
```

Sbozzolo replied:
I changed a little bit the cell structure to avoid having to redefine `datadir`

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -763,21 +781,55 @@
+    "# Normally you would just input the path or you would run the\n",
+    "# script from the output directory.\n",
+    "datadir=$(dirname $(./simfactory/bin/sim get-output-dir tov_ET))\n",
+    "./utils/Analysis/kuibit/examples/bins/plot_grid_var.py --datadir $datadir \\\n",
```
I'd add some text that explains what `x0` and `x1` are (lower left and uppper right corner?).

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -803,10 +859,99 @@
+    "# Undo reflection symmetry on the x axis\n",
+    "rho_xy_unif.reflection_symmetry_undo(dimension=0)\n",
+    "# Undo reflection symmetry on the y axis\n",
+    "rho_xy_unif.reflection_symmetry_undo(dimension=1)\n",
```
this is already in the released version? Only rot180 was missing? The notebook should work with the ET release version and not require main of kuibit.

Sbozzolo replied:
These functions were introduced in version `1.3.0` (ET_2021_11 comes with `1.3.3`). Undoing rotational symmetry is not implemented yet.

rhaas80 commented:
Ticket is here: https://bitbucket.org/einsteintoolkit/tickets/issues/2589/modernize-information-about-visualization

Sbozzolo commented:
I think I addressed the points, with the exception of `pip --force`. I guess that what you have to do is `yes | pip3 install -U --user kuibit==1.3.3`

rhaas80 commented:
there's apparently a `--no-input` option (which aborts though and does not continue) which is not quite what we need.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -101,7 +101,21 @@
+    "# We are going to install kuibit, a Python package to post-process Cactus simulations.\n",
+    "# We will install kuibit inside the Cactus directory. The main reason for this is to\n",
+    "# have a make easier to uninstall kuibit (you can just remove the Cactus folder). \n",
+    "os.environ[\"PYTHONUSERBASE\"] = os.path.join(os.environ[\"HOME\"], \"/Cactus\")\n",
```
I would say you should use either `os.path.join("Foo", "bar")` or `"foo" + "/bar"` ie the "/" in "/Cactus" is not needed. I did not suggest `os.path.join` mostly b/c later on in the notebook we already made the assumption that "/" is the path separator so using just string concatenation was fewer opaque function calls.

rhaas80 replied:
Actually... this fails since it sets `os.environ["PYTHONUSERBASE"]` to "/Cactus" (one assumes that `path.join` sees the initial "/" and takes it to be an absolute path). :-(

rhaas80 replied:
And... it does not even suffice to set PYTHONUSERBASE after Python has started. One needs to force it to reread the env variable, and this would have to happen _after_ `pip3 install kuibit` since Python checks if PYTHONUSERBASE contains a correct site-packages structure before adding it to `sys.path`. :-( 

Note that the same issue exists with not setting PYTHONUSERBASE at all. Namely this:  
```
mv ~/.local ~/\_local.bak  
python3 -c 'import os;os.system("pip3 install kuibit==1.3.3");import kuibit'  
```  
fails since at the time Python starts there is no `~/.local`. A second:  
```  
python3 -c 'import os;os.system("pip3 install kuibit==1.3.3");import kuibit'  
```  
succeeds since at that point `pip3` has created the expected directories.  
  
So: needs work, in all cases.

An ugly way to make this work is have a cell:
```
import site
site.USER_SITE = None
site.USER_BASE = None
site.main()
```
after running `pip3` which will force `site` to reinterpret `PYTHONUSERBASE` at a point where the looked for directories are present.

Sbozzolo replied:
Great catch!

It was working on my computer because I already have kuibit system-wide.

I added a small function to reload `PYTHONUSERBASE`. It should be good now.

rhaas80 replied:
Actually, this came up in today's ET call. Since we are already including a long list of `apt-get install` like commands to set up things for the ET, maybe instead of running `pip install` here, one could add `apt-get install kuibit` where possible and `pip install --user` otherwise.
Myself I would still want to add a `PYTHONUSERBASE` or so, so that things can be removed (since pip is not very good at cleaning up dependencies after a `pip uninstall`) but outside of the notebook that is harder to do. 
@stevenrbrandt also wanted to take a look and has suggestions as well.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -753,8 +779,35 @@
-    "# show the most recent segment directory that Cactus stored its output files in\n",
-    "./simfactory/bin/sim get-output-dir tov_ET"
+    "# See comment before\n",
+    "export PYTHONUSERBASE=\"$HOME/Cactus\"\n",
```
only needed once in a real shell, and for the notebook the only thing that (should) matters really is the `os.environ["PYTHONUSERBASE"]` bit at the beginning.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -763,21 +816,29 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "Image(filename=datadir+\"/rho_xy.png\")"
```
inconsistent use of `os.path.join` and "foo" + "/bar".

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -803,10 +868,99 @@
-    "plt.xlabel(r'$t$ [$M_{\\odot}$]');\n",
-    "plt.ylabel(r'$\\rho_c / \\rho_c(0)$');\n",
-    "plt.legend();"
+    "sim = SimDir(outdir)\n",
```
\`outdir\` seems to be undefined at this point. At least a search in the diff does not find it on the RHS of the diff (only on the LHS where it is removed). It is also not found when I look for `outdir` in the actual notebook.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -101,7 +101,21 @@
+    "    # Ask simfactory where the latest output was writted\n",
+    "    datadir = os.popen(\"./simfactory/bin/sim get-output-dir tov_ET\").read().rstrip(\"\\\\n\")\n",
+    "    # Remove the /output-XXX part\n",
+    "    return os.path.split(datadir)[0]"
```
awfully specific. Maybe better have it:  
\`\`\`python  
def get\_datadir(simname):  
   # Ask simfactory where the latest output was writted  
   datadir = os.popen(f"./simfactory/bin/sim get-output-dir {simname}").read().rstrip("\\\\n")  
   # Remove the /output-XXX part  
   return os.path.split(datadir)\[0\]  
\`\`\`  
to have a slighlty more generic function that could be used in other tutorials as well.

Sbozzolo replied:
I updated the code, but I find this function extremely ugly. Does simfactory have any API that we could use for this? What I mean is: would it be possible to import simfactory as a module and get information from there?

rhaas80 replied:
There is an API but it is not something that is really exposed. The only code that I know of using it is the `distribute` utility in `simfactory/bin` and it has to do a lot of magic to make things work.

There's also no Python function that would return the string that `get-output-dir` outputs. If you are interested, the function that handles that command is `command_get_output_dir` in `lib/sim-manage.py` which is the module to import.

stevenrbrandt replied:
I am not a fan of messing with PYTHONUSERBASE. It's a complication I'd rather not see. Just pip install?

Sbozzolo replied:
While I agree that it would be nice to keep things clean in users' environments, I am of the same opinion as @stevenrbrandt .

My take is that the more "infrastructure" functions we add, the harder it becomes to understand how to really use these tools.

rhaas80 replied:
With multiple different opinions we usually go by majority vote, so please go ahead with pip.

rhaas80 commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -763,21 +816,29 @@
-    "# outdir = outdir[0]\n",
-    "import os\n",
-    "outdir = os.popen(\"./simfactory/bin/sim get-output-dir tov_ET\").read().rstrip(\"\\n\")\n",
+    "from IPython.display import Image\n",
```
i may be better to run each plot command and then immediately show the produced file rather than first run two commands then show two files. This keeps the connection between command at output closer.

Sbozzolo replied:
I agree, but creating the two images in one bash cell keeps the code simpler.

Sbozzolo commented:
I updated the tutorial to properly reload `PYTHONUSERBASE`. I also removed the link to the video (which would be outdated). 

I think I took care of every requested change.

Sbozzolo commented:
@rhaas80 please review :)

stevenrbrandt commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -101,7 +101,32 @@
     "os.environ[\"OMPI_MCA_hwloc_base_binding_policy\"] = \"none\"\n",
     "os.environ[\"OMPI_MCA_btl_vader_single_copy_mechanism\"] = \"none\"\n",
-    "import scrolldown"
+    "#import scrolldown\n",
```
I would like to keep scrolldown. It could be installed with pip, or a try/except could be added.

Sbozzolo replied:
I will add a try/except.

stevenrbrandt commented on tutorial-server/notebooks/CactusTutorial.ipynb:
```diff
@@ -738,12 +763,24 @@
    "outputs": [],
    "source": [
-    "# This cell enables inline plotting in the notebook\n",
-    "%matplotlib inline\n",
```
Why are we getting rid of %matplotlib inline?

Sbozzolo replied:
We already have `%matplitlib inline` in another cell, and this achieves no effects, except reloading matplotlib.
